# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [55]:
#Import your libraries
import numpy as np
import pandas as pd

# Introduction

In this lab, we will use two datasets. Both datasets contain variables that describe apps from the Google Play Store. We will use our knowledge in feature extraction to process these datasets and prepare them for the use of a ML algorithm.

# Challenge 1 - Loading and Extracting Features from the First Dataset

#### In this challenge, our goals are: 

* Exploring the dataset.
* Identify the columns with missing values.
* Either replacing the missing values in each column or drop the columns.
* Conver each column to the appropriate type.

#### The first dataset contains different information describing the apps. 

Load the dataset into the variable `google_play` in the cell below. The dataset is in the file `googleplaystore.csv`

In [56]:
# Your code here:
google_play = pd.read_csv("../../data/googleplaystore.csv")
google_play["Price"]

0        0
1        0
2        0
3        0
4        0
        ..
10836    0
10837    0
10838    0
10839    0
10840    0
Name: Price, Length: 10841, dtype: object

#### Examine all variables and their types in the following cell

In [57]:
# Your code here:
google_play.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


#### Since this dataset only contains one numeric column, let's skip the `describe()` function and look at the first 5 rows using the `head()` function

In [58]:
# Your code here:
google_play.head(5)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

#### We can see that there are a few columns that could be coerced to numeric.

Start with the reviews column. We can evaluate what value is causing this column to be of object type finding the non-numeric values in this column. To do this, we recall the `to_numeric()` function. With this function, we are able to coerce all non-numeric data to null. We can then use the `isnull()` function to subset our dataframe using the True/False column that this function generates.

In the cell below, transform the Reviews column to numeric and assign this new column to the variable `Reviews_numeric`. Make sure to coerce the errors.

In [59]:
# Your code here:
Reviews_numeric = google_play["Reviews"].apply(pd.to_numeric, errors='coerce')


Reviews_numeric

0           159.0
1           967.0
2         87510.0
3        215644.0
4           967.0
           ...   
10836        38.0
10837         4.0
10838         3.0
10839       114.0
10840    398307.0
Name: Reviews, Length: 10841, dtype: float64

Next, create a column containing True/False values using the `isnull()` function. Assign this column to the `Reviews_isnull` variable.

In [60]:
# Your code here:
Reviews_isnull = Reviews_numeric.isnull()

Reviews_isnull

0        False
1        False
2        False
3        False
4        False
         ...  
10836    False
10837    False
10838    False
10839    False
10840    False
Name: Reviews, Length: 10841, dtype: bool

Finally, subset the `google_play` with `Reviews_isnull`. This should give you all the rows that contain non-numeric characters.

Your output should look like:

![Reviews_bool.png](../images/reviews-bool.png)

In [61]:
# Your code here:
missing_values = google_play[Reviews_isnull]
missing_values


App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

#### We see that Google Play is using a shorthand for millions. 

Let's write a function to transform this data.

Steps:

1. Create a function that returns the correct numeric values of *Reviews*.
1. Define a test string with `M` in the last character.
1. Test your function with the test string. Make sure your function works correctly. If not, modify your functions and test again.

In [62]:
# Your code here
import re

def convert_string_to_numeric(s):
    
    if s[-1] == 'M':
    
        numeric = s.replace('.', '')
        numeric = numeric.replace("M","")
        numeric = int(numeric)*100000
    else:
        numeric = s

    return float(numeric)

test_string = '4.0M'

convert_string_to_numeric(test_string) == 4000000

True

In [63]:
test_string

'4.0M'

The last step is to apply the function to the `Reviews` column in the following cell:

In [64]:
# Your code here:
#google_play.Reviews = austin_fixed.Date.apply(lambda x: dt.strptime(x, '%Y-%m-%d').toordinal())
#google_play.Reviews = google_play.Reviews.apply(lambda x : convert_string_to_numeric(x))

google_play["Reviews"] = google_play.Reviews.apply(convert_string_to_numeric)

Check the non-numeric `Reviews` row again. It should have been fixed now and you should see:

![Reviews_bool_fixed.png](../images/reviews-bool-fixed.png)

In [65]:
# Your code here
google_play[Reviews_isnull]

App Category  Rating    Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0  3000000.0   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

Also check the variable types of `google_play`. The `Reviews` column should be a `float64` type now.

In [66]:
# Your code here:
google_play.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  float64
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


#### The next column we will look at is `Size`. We start by looking at all unique values in `Size`:

*Hint: use `unique()` ([documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.unique.html))*.

In [67]:
# Your code here:
google_play.Size.unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

You should have seen lots of unique values of the app sizes.

#### While we can convert most of the `Size` values to numeric in the same way we converted the `Reviews` values, there is one value that is impossible to convert.

What is that badass value? Enter it in the next cell and calculate the proportion of its occurence to the total number of records of `google_play`.

In [68]:
# Your code here:
#len(google_play.loc[google_play['Size'] == 'Varies with device'])/len(google_play)
#



criteria = google_play['Size'] == 'Varies with device'

len(google_play[criteria])/len(google_play)

0.15635089013928605

#### While this column may be useful for other types of analysis, we opt to drop it from our dataset. 

There are two reasons. First, the majority of the data are ordinal but a sizeable proportion are missing because we cannot convert them to numerical values. Ordinal data are both numerical and categorical, and they usually can be ranked (e.g. 82k is smaller than 91M). In contrast, non-ordinal categorical data such as blood type and eye color cannot be ranked. The second reason is as a categorical column, it has too many unique values to produce meaningful insights. Therefore, in our case the simplest strategy would be to drop the column.

Drop the column in the cell below (use `inplace=True`)

In [69]:
# Your code here:
google_play.drop("Size", inplace=True,axis=1)

#### Now let's look at how many missing values are in each column. 

This will give us an idea of whether we should come up with a missing data strategy or give up on the column all together. In the next column, find the number of missing values in each column: 

*Hint: use the `isna()` and `sum()` functions.*

In [70]:
# Your code here:
google_play.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

You should find the column with the most missing values is now `Rating`.

#### What is the proportion of the missing values in `Rating` to the total number of records?

Enter your answer in the cell below.

In [71]:
# Your code here:
google_play["Rating"].isna().sum()/len(google_play)

0.13596531685268887

A sizeable proportion of the `Rating` column is missing. A few other columns also contain several missing values.

#### We opt to preserve these columns and remove the rows containing missing data.

In particular, we don't want to drop the `Rating` column because:

* It is one of the most important columns in our dataset. 

* Since the dataset is not a time series, the loss of these rows will not have a negative impact on our ability to analyze the data. It will, however, cause us to lose some meaningful observations. But the loss is limited compared to the gain we receive by preserving these columns.

In the cell below, remove all rows containing at least one missing value. Use the `dropna()` function ([documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dropna.html)). Assign the new dataframe to the variable `google_missing_removed`.

In [99]:
# Your code here:

google_missing_removed = google_play.dropna()
google_missing_removed

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating   Reviews     Installs  Type Price Content Rating  \
0         4.1     159.0      10,000+  Free     0       Everyone   
1         3.9     967.0     500,000+  Free     0       Everyone   
2         4.7   87510.0   5,000,000+  Free     0       Everyone   
3         4.5  215644.0  50,000,000+  Free     0           Teen   
4         4.3     967.0     100,000+  Free     0       Everyone   
...       ...       ...          ...   ...   ...            ...   
10834     4.0       7.0         500+  Free     0       Everyone   
10836     4.5      38.0       5,000+  Free     0       Everyone   
10837     5.0       4.0         100+  Free     0       Everyone   
10839     4.5     114.0       1,000+  Free     0     Mature 17+   
10840     4.5  398307.0  10,000,000+  Free     0       Everyone   

                          Genres      Last Updated         Current Ver  \
0                   Art & Design   January 7, 2018               1.0.0   
1      Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                   Art & Design    August 1, 2018               1.2.4   
3                   Art & Design      June 8, 2018  Varies with device   
4        Art & Design;Creativity     June 20, 2018                 1.1   
...                          ...               ...                 ...   
10834                  Education     June 18, 2017               1.0.0   
10836                  Education     July 25, 2017                1.48   
10837                  Education      July 6, 2018                 1.0   
10839          Books & Reference  January 19, 2015  Varies with device   
10840                  Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
0            4.0.3 and up  
1            4.0.3 and up  
2            4.0.3 and up  
3              4.2 and up  
4              4.4 and up  
...                   ...  
10834          4.1 and up  
10836          4.1 and up  
10837          4.1 and up  
10839  Varies with device  
10840  Varies with device  

[9360 rows x 12 columns]

From now on, we use the `google_missing_removed` variable instead of `google_play`.

#### Next, we look at the `Last Updated` column.

The `Last Updated` column seems to contain a date, though it is classified as an object type. Let's convert this column using the `pd.to_datetime` function ([documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)).

In [100]:
# Your code here:
google_missing_removed['Last Updated'] = pd.to_datetime(google_missing_removed['Last Updated'])
google_missing_removed['Last Updated']

C:\Users\Usuario\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


0       2018-01-07
1       2018-01-15
2       2018-08-01
3       2018-06-08
4       2018-06-20
           ...    
10834   2017-06-18
10836   2017-07-25
10837   2018-07-06
10839   2015-01-19
10840   2018-07-25
Name: Last Updated, Length: 9360, dtype: datetime64[ns]

#### The last column we will transform is `Price`. 

We start by looking at the unique values of this column.

In [101]:
# Your code here:
google_missing_removed['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$7.99', '$5.99', '$2.99', '$3.49',
       '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49', '$10.00',
       '$24.99', '$11.99', '$79.99', '$16.99', '$14.99', '$29.99',
       '$12.99', '$2.49', '$10.99', '$1.50', '$19.99', '$15.99', '$33.99',
       '$39.99', '$3.95', '$4.49', '$1.70', '$8.99', '$1.49', '$3.88',
       '$399.99', '$17.99', '$400.00', '$3.02', '$1.76', '$4.84', '$4.77',
       '$1.61', '$2.50', '$1.59', '$6.49', '$1.29', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$8.49', '$1.75', '$14.00', '$2.00',
       '$3.08', '$2.59', '$19.40', '$3.90', '$4.59', '$15.46', '$3.04',
       '$13.99', '$4.29', '$3.28', '$4.60', '$1.00', '$2.95', '$2.90',
       '$1.97', '$2.56', '$1.20'], dtype=object)

Since all prices are ordinal data without exceptions, we can tranform this column by removing the dollar sign and converting to numeric. We can create a new column called `Price Numerical` and drop the original column.

We will achieve our goal in three steps. Follow the instructions of each step below.

#### First we remove the dollar sign. Do this in the next cell by applying the `str.replace` function to the column to replace `$` with an empty string (`''`).

In [102]:
google_missing_removed

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating   Reviews     Installs  Type Price Content Rating  \
0         4.1     159.0      10,000+  Free     0       Everyone   
1         3.9     967.0     500,000+  Free     0       Everyone   
2         4.7   87510.0   5,000,000+  Free     0       Everyone   
3         4.5  215644.0  50,000,000+  Free     0           Teen   
4         4.3     967.0     100,000+  Free     0       Everyone   
...       ...       ...          ...   ...   ...            ...   
10834     4.0       7.0         500+  Free     0       Everyone   
10836     4.5      38.0       5,000+  Free     0       Everyone   
10837     5.0       4.0         100+  Free     0       Everyone   
10839     4.5     114.0       1,000+  Free     0     Mature 17+   
10840     4.5  398307.0  10,000,000+  Free     0       Everyone   

                          Genres Last Updated         Current Ver  \
0                   Art & Design   2018-01-07               1.0.0   
1      Art & Design;Pretend Play   2018-01-15               2.0.0   
2                   Art & Design   2018-08-01               1.2.4   
3                   Art & Design   2018-06-08  Varies with device   
4        Art & Design;Creativity   2018-06-20                 1.1   
...                          ...          ...                 ...   
10834                  Education   2017-06-18               1.0.0   
10836                  Education   2017-07-25                1.48   
10837                  Education   2018-07-06                 1.0   
10839          Books & Reference   2015-01-19  Varies with device   
10840                  Lifestyle   2018-07-25  Varies with device   

              Android Ver  
0            4.0.3 and up  
1            4.0.3 and up  
2            4.0.3 and up  
3              4.2 and up  
4              4.4 and up  
...                   ...  
10834          4.1 and up  
10836          4.1 and up  
10837          4.1 and up  
10839  Varies with device  
10840  Varies with device  

[9360 rows x 12 columns]

In [103]:
#lambda x: True if x % 2 == 0 else False


google_missing_removed["Price_Numerical"] = google_missing_removed.Price.apply(lambda x :  x.replace('$', '') if x != 0 else x)

C:\Users\Usuario\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


# Your code here:
def convert_string_to_numeric2(s):
    
    if s != 0:
    
        numeric = s.replace('$', '')
    else:
        numeric = s

    return float(numeric)

test_string = "$4.99"
google_missing_removed["Price_Numerical"] = google_missing_removed.Price.apply(convert_string_to_numeric2)
google_missing_removed.drop("Price", inplace=True,axis=1)
google_missing_removed

In [104]:
google_missing_removed["Price_Numerical"].unique()


array(['0', '4.99', '3.99', '6.99', '7.99', '5.99', '2.99', '3.49',
       '1.99', '9.99', '7.49', '0.99', '9.00', '5.49', '10.00', '24.99',
       '11.99', '79.99', '16.99', '14.99', '29.99', '12.99', '2.49',
       '10.99', '1.50', '19.99', '15.99', '33.99', '39.99', '3.95',
       '4.49', '1.70', '8.99', '1.49', '3.88', '399.99', '17.99',
       '400.00', '3.02', '1.76', '4.84', '4.77', '1.61', '2.50', '1.59',
       '6.49', '1.29', '299.99', '379.99', '37.99', '18.99', '389.99',
       '8.49', '1.75', '14.00', '2.00', '3.08', '2.59', '19.40', '3.90',
       '4.59', '15.46', '3.04', '13.99', '4.29', '3.28', '4.60', '1.00',
       '2.95', '2.90', '1.97', '2.56', '1.20'], dtype=object)

#### Second step, coerce the `Price Numerical` column to numeric.

In [113]:
# Your code here:
google_missing_removed["Price_Numerical"] = google_missing_removed["Price_Numerical"].apply(pd.to_numeric, errors='coerce')

C:\Users\Usuario\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


**Finally, drop the original `Price` column.**

In [114]:
# Your code here:
google_missing_removed.drop("Price", inplace=True,axis=1)


KeyError: "['Price'] not found in axis"

In [116]:
google_missing_removed["Price_Numerical"].unique()

array([  0.  ,   4.99,   3.99,   6.99,   7.99,   5.99,   2.99,   3.49,
         1.99,   9.99,   7.49,   0.99,   9.  ,   5.49,  10.  ,  24.99,
        11.99,  79.99,  16.99,  14.99,  29.99,  12.99,   2.49,  10.99,
         1.5 ,  19.99,  15.99,  33.99,  39.99,   3.95,   4.49,   1.7 ,
         8.99,   1.49,   3.88, 399.99,  17.99, 400.  ,   3.02,   1.76,
         4.84,   4.77,   1.61,   2.5 ,   1.59,   6.49,   1.29, 299.99,
       379.99,  37.99,  18.99, 389.99,   8.49,   1.75,  14.  ,   2.  ,
         3.08,   2.59,  19.4 ,   3.9 ,   4.59,  15.46,   3.04,  13.99,
         4.29,   3.28,   4.6 ,   1.  ,   2.95,   2.9 ,   1.97,   2.56,
         1.2 ])

Now check the variable types of `google_missing_removed`. Make sure:

* `Size` and `Price` columns have been removed.
* `Rating`, `Reviews`, and `Price Numerical` have the type of `float64`.
* `Last Updated` has the type of `datetime64`.

In [121]:
# Your code here
google_missing_removed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9360 entries, 0 to 10840
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   App              9360 non-null   object        
 1   Category         9360 non-null   object        
 2   Rating           9360 non-null   float64       
 3   Reviews          9360 non-null   float64       
 4   Installs         9360 non-null   object        
 5   Type             9360 non-null   object        
 6   Content Rating   9360 non-null   object        
 7   Genres           9360 non-null   object        
 8   Last Updated     9360 non-null   datetime64[ns]
 9   Current Ver      9360 non-null   object        
 10  Android Ver      9360 non-null   object        
 11  Price_Numerical  9360 non-null   float64       
dtypes: datetime64[ns](1), float64(3), object(8)
memory usage: 950.6+ KB


# Challenge 2 - Loading and Extracting Features from the Second Dataset

Load the second dataset to the variable `google_reviews`. The data is in the file `googleplaystore_user_reviews.csv`.

In [119]:
# Your code here:
google_reviews = pd.read_csv("../../data/googleplaystore_user_reviews.csv")
google_reviews

App  \
0            10 Best Foods for You   
1            10 Best Foods for You   
2            10 Best Foods for You   
3            10 Best Foods for You   
4            10 Best Foods for You   
...                            ...   
64290  Houzz Interior Design Ideas   
64291  Houzz Interior Design Ideas   
64292  Houzz Interior Design Ideas   
64293  Houzz Interior Design Ideas   
64294  Houzz Interior Design Ideas   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
2                                                    NaN       NaN   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
...                                                  ...       ...   
64290                                                NaN       NaN   
64291                                                NaN       NaN   
64292                                                NaN       NaN   
64293                                                NaN       NaN   
64294                                                NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                    1.00                0.533333  
1                    0.25                0.288462  
2                     NaN                     NaN  
3                    0.40                0.875000  
4                    1.00                0.300000  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[64295 rows x 5 columns]

#### This dataset contains the top 100 reviews for each app. 

Let's examine this dataset using the `head` function

In [123]:
# Your code here
google_reviews.head(30)

App                                  Translated_Review  \
0   10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1   10 Best Foods for You    This help eating healthy exercise regular basis   
2   10 Best Foods for You                                                NaN   
3   10 Best Foods for You         Works great especially going grocery store   
4   10 Best Foods for You                                       Best idea us   
5   10 Best Foods for You                                           Best way   
6   10 Best Foods for You                                            Amazing   
7   10 Best Foods for You                                                NaN   
8   10 Best Foods for You                               Looking forward app,   
9   10 Best Foods for You              It helpful site ! It help foods get !   
10  10 Best Foods for You                                          good you.   
11  10 Best Foods for You  Useful information The amount spelling errors ...   
12  10 Best Foods for You  Thank you! Great app!! Add arthritis, eyes, im...   
13  10 Best Foods for You  Greatest ever Completely awesome maintain heal...   
14  10 Best Foods for You  Good health...... Good health first priority.....   
15  10 Best Foods for You                                                NaN   
16  10 Best Foods for You  Health It's important world either life . thin...   
17  10 Best Foods for You  Mrs sunita bhati I thankful developers,to make...   
18  10 Best Foods for You  Very Useful in diabetes age 30. I need control...   
19  10 Best Foods for You                                 One greatest apps.   
20  10 Best Foods for You                                          good nice   
21  10 Best Foods for You                              Healthy Really helped   
22  10 Best Foods for You                                         God health   
23  10 Best Foods for You  HEALTH SHOULD ALWAYS BE TOP PRIORITY. !!. ON M...   
24  10 Best Foods for You                              An excellent A useful   
25  10 Best Foods for You                  I found lot wealth form health...   
26  10 Best Foods for You                         Because I found important.   
27  10 Best Foods for You                                     Healthy Eating   
28  10 Best Foods for You                              Very good Simply good   
29  10 Best Foods for You                                        On test....   

   Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0   Positive            1.000000                0.533333  
1   Positive            0.250000                0.288462  
2        NaN                 NaN                     NaN  
3   Positive            0.400000                0.875000  
4   Positive            1.000000                0.300000  
5   Positive            1.000000                0.300000  
6   Positive            0.600000                0.900000  
7        NaN                 NaN                     NaN  
8    Neutral            0.000000                0.000000  
9    Neutral            0.000000                0.000000  
10  Positive            0.700000                0.600000  
11  Positive            0.200000                0.100000  
12  Positive            0.750000                0.875000  
13  Positive            0.992188                0.866667  
14  Positive            0.550000                0.511111  
15       NaN                 NaN                     NaN  
16  Positive            0.450000                1.000000  
17  Positive            0.600000                0.666667  
18  Positive            0.295000                0.100000  
19  Positive            1.000000                1.000000  
20  Positive            0.650000                0.800000  
21  Positive            0.350000                0.350000  
22   Neutral            0.000000                0.000000  
23  Positive            0.781250                0.500000  
24  Positive            0.650000                0.500000  
25   Ne

#### The main piece of information we would like to extract from this dataset is the proportion of positive reviews of each app. 

Columns like `Sentiment_Polarity` and `Sentiment_Subjectivity` are not to our interests because we have no clue how to use them. We do not care about `Translated_Review` because natural language processing is too complex for us at present (in fact the `Sentiment`, `Sentiment_Polarity`, and `Sentiment_Subjectivity` columns are derived from `Translated_Review` the data scientists). 

What we care about in this challenge is `Sentiment`. To be more precise, we care about **what is the proportion of *Positive* sentiment of each app**. This will require us to aggregate the `Sentiment` data by `App` in order to calculate the proportions.

Now that you are clear about what we are trying to achieve, follow the steps below that will walk you through towards our goal.

#### Our first step will be to remove all rows with missing sentiment. 

In the next cell, drop all rows with missing data using the `dropna()` function and assign this new dataframe to `review_missing_removed`.

In [196]:
# Your code here:
review_missing_removed = google_reviews.dropna()
review_missing_removed

App  \
0               10 Best Foods for You   
1               10 Best Foods for You   
3               10 Best Foods for You   
4               10 Best Foods for You   
5               10 Best Foods for You   
...                               ...   
64222  Housing-Real Estate & Property   
64223  Housing-Real Estate & Property   
64226  Housing-Real Estate & Property   
64227  Housing-Real Estate & Property   
64230  Housing-Real Estate & Property   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
5                                               Best way  Positive   
...                                                  ...       ...   
64222  Most ads older many agents ..not much owner po...  Positive   
64223  If photos posted portal load, fit purpose. I'm...  Positive   
64226  Dumb app, I wanted post property rent give opt...  Negative   
64227  I property business got link SMS happy perform...  Positive   
64230  Useless app, I searched flats kondapur, Hydera...  Negative   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                1.000000                0.533333  
1                0.250000                0.288462  
3                0.400000                0.875000  
4                1.000000                0.300000  
5                1.000000                0.300000  
...                   ...                     ...  
64222            0.173333                0.486667  
64223            0.225000                0.447222  
64226           -0.287500                0.250000  
64227            0.800000                1.000000  
64230           -0.316667                0.400000  

[37427 rows x 5 columns]

#### Now, use the `value_counts()` function ([documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)) to get a sense on how many apps are in this dataset and their review counts.

In [197]:
# Your code here:
len(review_missing_removed.App.value_counts())
review_missing_removed.App.value_counts()

Bowmasters                                           312
Helix Jump                                           273
Angry Birds Classic                                  273
Calorie Counter - MyFitnessPal                       254
Duolingo: Learn Languages Free                       240
                                                    ... 
CallApp: Caller ID, Blocker & Phone Call Recorder      1
All-in-One Mahjong 3 FREE                              1
Bed Time Fan - White Noise Sleep Sounds                1
Draw A Stickman                                        1
Free Live Talk-Video Call                              1
Name: App, Length: 865, dtype: int64

#### Now the tough part comes. Let's plan how we will achieve our goal:

1. We will count the number of reviews that contain *Positive* in the `Sentiment` column.

1. We will create a new dataframe to contain the `App` name, the number of positive reviews, and the total number of reviews of each app.

1. We will then loop the new dataframe to calculate the postivie review portion of each app.

#### Step 1: Count the number of positive reviews.

In the following cell, write a function that takes a column and returns the number of times *Positive* appears in the column. 

*Hint: One option is to use the `np.where()` function ([documentation](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.where.html)).*

In [201]:
count = 0
for i in review_missing_removed["Sentiment"]:
    if i == 'Positive':
        count +=1

In [202]:
count

23998

In [206]:
# Your code below    
    #criteria = x == 'Positive'

    #positive_count = len(x[criteria])
import numpy as np

def positive_function(s):
    count = 0
    for i in s:
        if i == 'Positive':
            count +=1

    """
    Count how many times the string `Positive` appears in a column (exact string match).
    
    Args:
        x: data column
    
    Returns:
        The number of occurrences of `Positive` in the column data.
    """
    return count

In [207]:
positive_function(review_missing_removed["Sentiment"])

23998

In [208]:
review_missing_removed

App  \
0               10 Best Foods for You   
1               10 Best Foods for You   
3               10 Best Foods for You   
4               10 Best Foods for You   
5               10 Best Foods for You   
...                               ...   
64222  Housing-Real Estate & Property   
64223  Housing-Real Estate & Property   
64226  Housing-Real Estate & Property   
64227  Housing-Real Estate & Property   
64230  Housing-Real Estate & Property   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
5                                               Best way  Positive   
...                                                  ...       ...   
64222  Most ads older many agents ..not much owner po...  Positive   
64223  If photos posted portal load, fit purpose. I'm...  Positive   
64226  Dumb app, I wanted post property rent give opt...  Negative   
64227  I property business got link SMS happy perform...  Positive   
64230  Useless app, I searched flats kondapur, Hydera...  Negative   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                1.000000                0.533333  
1                0.250000                0.288462  
3                0.400000                0.875000  
4                1.000000                0.300000  
5                1.000000                0.300000  
...                   ...                     ...  
64222            0.173333                0.486667  
64223            0.225000                0.447222  
64226           -0.287500                0.250000  
64227            0.800000                1.000000  
64230           -0.316667                0.400000  

[37427 rows x 5 columns]

#### Step 2: Create a new dataframe to contain the `App` name, the number of positive reviews, and the total number of reviews of each app

We will group `review_missing_removed` by the `App` column, then aggregate the grouped dataframe on the number of positive reviews and the total review counts of each app. The result will be assigned to a new variable `google_agg`. Here is the ([documentation on how to achieve it](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.core.groupby.DataFrameGroupBy.agg.html)). Take a moment or two to read the documentation and google examples because it is pretty complex.

When you obtain `google_agg`, check its values to make sure it has an `App` column as its index as well as a `Positive` column and a `Total` column. Your output should look like:

![Positive Reviews Agg](../images/positive-review-agg.png)

*Hint: Use `positive_function` you created earlier as part of the param passed to the `agg()` function in order to aggregate the number of positive reviews.*

#### Bonus:

As of Pandas v0.23.4, you may opt to supply an array or an object to `agg()`. If you use the array param, you'll need to rename the columns so that their names are `Positive` and `Total`. Using the object param will allow you to create the aggregated columns with the desirable names without renaming them. However, you will probably encounter a warning indicating supplying an object to `agg()` will become outdated. It's up to you which way you will use. Try both ways out. Any way is fine as long as it works.

In [210]:
# Your code here:
google_agg = review_missing_removed.groupby(['App']).agg({'Sentiment': [positive_function, "count"]})


google_agg

Sentiment      
                                                   positive_function count
App                                                                       
10 Best Foods for You                                            162   194
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                                  31    40
11st                                                              23    39
1800 Contacts - Lens Store                                        64    80
1LINE – One Line with One Touch                                   27    38
...                                                              ...   ...
Hotels.com: Book Hotel Rooms & Find Vacation Deals                39    68
Hotspot Shield Free VPN Proxy & Wi-Fi Security                    17    34
Hotstar                                                           14    32
Hotwire Hotel & Car Rental App                                    16    33
Housing-Real Estate & Property                                     8    21

[865 rows x 2 columns]

Print the first 5 rows of `google_agg` to check it.

In [227]:
# Your code here
google_agg.head(5)

Positive_Ratio
App                                             
10 Best Foods for You                   0.835052
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室        0.775000
11st                                    0.589744
1800 Contacts - Lens Store              0.800000
1LINE – One Line with One Touch         0.710526

#### Add a derived column to `google_agg` that is the ratio of the `Positive` and the `Total` columns. Call this column `Positive Ratio`. 

Make sure to account for the case where the denominator is zero using the `np.where()` function.

In [223]:
# Your code here:
#google_agg["Positive_Ratio"] = google_agg["positive_function"]/google_agg["count"]


In [224]:
google_agg

Positive_Ratio
App                                                               
10 Best Foods for You                                     0.835052
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                          0.775000
11st                                                      0.589744
1800 Contacts - Lens Store                                0.800000
1LINE – One Line with One Touch                           0.710526
...                                                            ...
Hotels.com: Book Hotel Rooms & Find Vacation Deals        0.573529
Hotspot Shield Free VPN Proxy & Wi-Fi Security            0.500000
Hotstar                                                   0.437500
Hotwire Hotel & Car Rental App                            0.484848
Housing-Real Estate & Property                            0.380952

[865 rows x 1 columns]

#### Now drop the `Positive` and `Total` columns. Do this with `inplace=True`.

In [33]:
# Your code here:
google_agg.drop("positive_function",axis=1,inplace=True)
google_agg.drop("count",axis=1,inplace=True)

Print the first 5 rows of `google_agg`. Your output should look like:

![Positive Reviews Agg](../images/positive-review-ratio.png)

In [228]:
# Your code here:
google_agg.head(5)

Positive_Ratio
App                                             
10 Best Foods for You                   0.835052
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室        0.775000
11st                                    0.589744
1800 Contacts - Lens Store              0.800000
1LINE – One Line with One Touch         0.710526

# Challenge 3 - Join the Dataframes

In this part of the lab, we will join the two dataframes and obtain a dataframe that contains features we can use in our ML algorithm.

In the next cell, join the `google_missing_removed` dataframe with the `google_agg` dataframe on the `App` column. Assign this dataframe to the variable `google`.

In [233]:
# Your code here:
google = google_missing_removed.merge(google_agg,left_on="App",right_on="App")

#### Let's look at the final result using the `head()` function. Your final product should look like:

![Final Product](../images/google-final-head.png)

In [234]:
# Your code here:
google.head(5)

App        Category  Rating  \
0                        Coloring book moana  ART_AND_DESIGN     3.9   
1                        Coloring book moana          FAMILY     3.9   
2                       Garden Coloring Book  ART_AND_DESIGN     4.4   
3              FlipaClip - Cartoon animation  ART_AND_DESIGN     4.3   
4  Boys Photo Editor - Six Pack & Men's Suit  ART_AND_DESIGN     4.1   

    Reviews    Installs  Type Content Rating                     Genres  \
0     967.0    500,000+  Free       Everyone  Art & Design;Pretend Play   
1     974.0    500,000+  Free       Everyone  Art & Design;Pretend Play   
2   13791.0  1,000,000+  Free       Everyone               Art & Design   
3  194216.0  5,000,000+  Free       Everyone               Art & Design   
4     654.0    100,000+  Free       Everyone               Art & Design   

  Last Updated Current Ver   Android Ver  Price_Numerical  Positive_Ratio  
0   2018-01-15       2.0.0  4.0.3 and up              0.0        0.590909  
1   2018-01-15       2.0.0  4.0.3 and up              0.0        0.590909  
2   2017-09-20       2.9.2    3.0 and up              0.0        0.711111  
3   2018-08-03       2.2.5  4.0.3 and up              0.0        1.000000  
4   2018-03-20         1.1  4.0.3 and up              0.0        0.605263

In [235]:
google

App            Category  Rating  \
0                           Coloring book moana      ART_AND_DESIGN     3.9   
1                           Coloring book moana              FAMILY     3.9   
2                          Garden Coloring Book      ART_AND_DESIGN     4.4   
3                 FlipaClip - Cartoon animation      ART_AND_DESIGN     4.3   
4     Boys Photo Editor - Six Pack & Men's Suit      ART_AND_DESIGN     4.1   
...                                         ...                 ...     ...   
1223  AD - Nieuws, Sport, Regio & Entertainment  NEWS_AND_MAGAZINES     3.7   
1224                                    AppLock               TOOLS     4.4   
1225                   Diary with lock password           LIFESTYLE     4.3   
1226             Avakin Life - 3D virtual world              FAMILY     4.4   
1227               A+ Gallery - Photos & Videos         PHOTOGRAPHY     4.5   

        Reviews      Installs  Type Content Rating                     Genres  \
0         967.0      500,000+  Free       Everyone  Art & Design;Pretend Play   
1         974.0      500,000+  Free       Everyone  Art & Design;Pretend Play   
2       13791.0    1,000,000+  Free       Everyone               Art & Design   
3      194216.0    5,000,000+  Free       Everyone               Art & Design   
4         654.0      100,000+  Free       Everyone               Art & Design   
...         ...           ...   ...            ...                        ...   
1223    10369.0    1,000,000+  Free       Everyone           News & Magazines   
1224  4931562.0  100,000,000+  Free       Everyone                      Tools   
1225   179139.0    5,000,000+  Free       Everyone                  Lifestyle   
1226  1275373.0   10,000,000+  Free           Teen               Role Playing   
1227   223941.0   10,000,000+  Free       Everyone                Photography   

     Last Updated         Current Ver         Android Ver  Price_Numerical  \
0      2018-01-15               2.0.0        4.0.3 and up              0.0   
1      2018-01-15               2.0.0        4.0.3 and up              0.0   
2      2017-09-20               2.9.2          3.0 and up              0.0   
3      2018-08-03               2.2.5        4.0.3 and up              0.0   
4      2018-03-20                 1.1        4.0.3 and up              0.0   
...           ...                 ...                 ...              ...   
1223   2018-07-28  Varies with device  Varies with device              0.0   
1224   2018-06-11  Varies with device  Varies with device              0.0   
1225   2017-11-01               3.5.3        4.0.3 and up              0.0   
1226   2018-07-03            1.021.10          4.1 and up              0.0   
1227   2018-08-06  Varies with device  Varies with device              0.0   

      Positive_Ratio  
0           0.590909  
1           0.590909  
2           0.711111  
3           1.000000  
4           0.605263  
...              ...  
1223        0.684211  
1224        0.461538  
1225        0.730769  
1226        0.753623  
1227        0.585366  

[1228 rows x 13 columns]